# vcf2fasta.pl
Converts a VCF file to a FASTA alignment provided a reference genome and a GFF file

Given a FASTA reference genome, a multi-sample VCF file and a GFF file, this script will generate FASTA alignments of any feature found in the GFF file; for instance, coding sequences (CDS). Gene names will be taken from the first id in field 9 of the GFF. It is important that the GFF file is sorted by position. The script will take diploid, phased, and/or haploid data.Â It is important to note that the script currently only takes very standard VCF formats; for instance, those generated by haplotype-based algorithms (HaplotypeCaller and FreeBayes), which include gaps and multi-nucleotide variants, will not work.

## Installation

    git clone https://github.com/santiagosnchez/vcf2fasta
    cd vcf2fasta
    chmod +x vcf2fasta.pl
    sudo cp vcf2fasta.pl /usr/local/bin

## Running the code

Use the `-h` flag for more details:

    perl vcf2fasta.pl -h
    Usage:
    perl vcf2fasta.pl -f <fasta-ref> -v <vcf-file> -g <gff-file> -e <gff-feature> [ --ref ] [ --phased ]
    
    option --ref will include the reference sequence.
    option --phase informs the program that the VCF is pased.
    defaults ignore the two options.
    
    examples:
    perl vcf2fasta.pl -f ref.fas -v snps.vcf -g annotation.gff -e CDS
    perl vcf2fasta.pl -f ref.fas -v snps.phased.vcf -g annotation.gff -e CDS --phased
    perl vcf2fasta.pl -f ref.fas -v snps.phased.vcf -g annotation.gff -e CDS --phased --ref

## Note: Working on a Python version...
